vmm/src/serial_manager.rs

```diff
@@ -0,0 +1,235 @@
+// Copyright © 2021 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#[cfg(target_arch = "aarch64")]
+use devices::legacy::Pl011;
+#[cfg(target_arch = "x86_64")]
+use devices::legacy::Serial;
+use std::fs::File;
+use std::io::Read;
+use std::os::unix::io::{AsRawFd, FromRawFd, RawFd};
+use std::sync::{Arc, Mutex};
+use std::{io, result};
+use thiserror::Error;
+use vmm_sys_util::terminal::Terminal;
+
+#[derive(Debug, Error)]
+pub enum Error {
+    /// Cannot clone File
+    #[error("Error cloning File: {0}")]
+    FileClone(#[source] io::Error),
+
+    /// Cannot create epoll context.
+    #[error("Error creating epoll context: {0}")]
+    Epoll(#[source] io::Error),
+
+    /// Cannot handle the VM STDIN stream
+    #[error("Error handling VM stdin: {0:?}")]
+    Stdin(#[source] vmm_sys_util::errno::Error),
+
+    /// Cannot handle the VM pty stream
+    #[error("Error handling VM pty: {0:?}")]
+    Pty(#[source] io::Error),
+
+    /// Cannot queue input to the serial device
+    #[error("Error queuing input to the serial device: {0}")]
+    QueueInput(#[source] vmm_sys_util::errno::Error),
+}
+pub type Result<T> = result::Result<T, Error>;
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+#[repr(u64)]
+pub enum EpollDispatch {
+    Stdin = 0,
+    SerialPty = 1,
+    Unknown,
+}
+
+impl From<u64> for EpollDispatch {
+    fn from(v: u64) -> Self {
+        use EpollDispatch::*;
+        match v {
+            0 => Stdin,
+            1 => SerialPty,
+            _ => Unknown,
+        }
+    }
+}
+
+pub struct EpollContext {
+    epoll_file: File,
+}
+
+impl EpollContext {
+    pub fn new() -> result::Result<EpollContext, io::Error> {
+        let epoll_fd = epoll::create(true)?;
+        // Use 'File' to enforce closing on 'epoll_fd'
+        let epoll_file = unsafe { File::from_raw_fd(epoll_fd) };
+
+        Ok(EpollContext { epoll_file })
+    }
+
+    pub fn add_stdin(&mut self) -> result::Result<(), io::Error> {
```

Comment:
I don't think you need a special case for stdin here it's got a FD number like any other file.